### ISSUE TITLE
A feeds B if B has intrinsic value of 1 or greater

### ISSUE THREAD

**Issue by DrHeinzDoofenshmirtz** [derrickcreamer/Hemlock#1]
Hi.

Awesome library.
I am exploring it at the moment, and I have a case that I can't quite figure out how to do.

I want to expose two "real" statuses A and B.

A should feed B, but only if B already has a value of 1 or greater.
Meaning, if B does not have a value, then I don't want to feed it from A.
A should not be able to give value to B just on its own.

Another bonus question:
Is it possible to make conditions for Feed that depends on values on other statuses?
At the moment seems that the rules need to be set before you can use values from the StatusTracker. And the Feeds extension's Func gives you the value from the status that feeds.

**Comment by derrickcreamer** [comment]
Hey, glad you found it interesting!

So, let's see... What do you want to happen if:
 - B already has a value of 2 by some means
 - A starts feeding B, let's say a value of 5, so B's total is 7
 - the original instance or feed that provided the value of 2 to B is removed?

Would you want the value of B to drop to zero in that case, or would you want it to get 'stuck' at the higher value as long as A is providing some value? I think I found a way to model this using four or five statuses (yikes! but at least it's possible). I'll describe how in the next comment.

(I've briefly considered adding a way to designate certain statuses as 'calculated', where they'd basically never participate in the usual relationships between statuses, and instead would have values calculated only when requested, based on the other statuses in this tracker -- I wonder if that would be useful for this case.)

**Comment by derrickcreamer** [comment]
So, this is a complicated relationship but I think I found a way. First I'll describe how to do the version that drops B back to zero if all non-A sources are removed, and then I'll describe how to modify it if you want the other version instead.

I'll define five statuses as follows:
 - OriginalA. This is the 'real' status for A. You'd add status instances to this one, and you'd check A's value by checking this one.
 - NaturalB. This is where you'd _add_ instances of B, but not where you'd check the final value.
 - FinalB. This is where you'd check the final value of B. (See rules below)
 - FromAToB. This one is never used directly, but has an important role to play in the calculation (see below).
 - C. This one is configured in a special way and is the key to making this work.

Here are the rules I'm imagining for this arrangement:
```
OriginalA feeds FromAToB
FromAToB feeds FinalB
NaturalB feeds FinalB
C suppresses FromAToB
NaturalB suppresses C
```
...but in order for this to work, the value of C must start at a positive value, and therefore you'd need to manually add an instance of C when each status tracker is created, before any A or B are added.

I think this would work as follows:
 - You can use OriginalA as normal, but by default C is true, and suppresses FromAToB, so A doesn't reach B yet.
 - As soon as NaturalB has value, C is suppressed, which means FromAToB is no longer suppressed, and the value of A is fed indirectly into FinalB (regardless of when it was added).
 - If NaturalB loses all of its value, C is no longer suppressed, which means that FromAToB becomes suppressed again, so the value of FinalB decreases.

Or, if you want it to work the other way, just combine NaturalB and FinalB into just B, making the rules like this:
```
OriginalA feeds FromAToB
FromAToB feeds B
C suppresses FromAToB
B suppresses C
```

Then, as soon as B has any value, C will be suppressed and let more value into B, after which B's value will only drop to zero if all instances of both A and B are removed.

Hope that helps! As for your bonus question, I'm curious about the use case - do you think it would be covered by the idea of separately-calculated values, or would it need to participate in the feeding of values directly? The idea of making statuses depend on multiple others is intriguing but also a bit intimidating. (I wonder if those would be best done as a special concept or not...)

**Comment by DrHeinzDoofenshmirtz** [comment]
Hi Derrick.

Thanks you for your very elaborate answer. I really appreciate it.
Those are some very good ideas!

I was thinking about the situation where A stops feeding B if "Natural B" has no value.

Your first first example should work, I think. I have never considered having separate statuses for Read and Write. I would have to be very disciplined to do that correctly. But it should work.
For the other example, wouldn't it be an infinite loop? B --> C --> FromAToB --> B
But as the first example it perfect for my use case, that is fine :)

And for the idea about calculated values. I think it would work.
At the moment, the issue is that when you set up the rules, you only have access to the value of the status you are making a rule for. And because the StatusTracker is created after all rules, you don't have access to the other values.
I could see any approach work, where you can set up a method/function/equation and define it using the values from the status.
I don't see how it would work with the Feed, Suppress, and so on system, but maybe you have an idea.
A calculated status sounds smart. It should probably be read-only to avoid issues. Otherwise I lso see it being very intimidating and confusing to create :)

**Comment by DrHeinzDoofenshmirtz** [comment]
Actually, if there was a way to have a custom function define how the value of a read-only status is calculated, and with access to the values of the different statuses, that would actually allow for a whole new way of defining the rules, and you can code all of it yourself. So for those cases where the normal setup is not sufficient, you can do anything you want basically.
That would be a nice addition if you ever get time and mental capacity for it :)

**Comment by DrHeinzDoofenshmirtz** [comment]
I tried your approach, but switching it around so that I wouldn't need to set the value of C at the beginning.

``` csharp
rules[Status.Shock].Feeds(StatusBookKeeping.ShockToElectrocuted);
rules[StatusBookKeeping.ShockToElectrocuted].Feeds(Status.ElectrocutedFinal);
// Shock --> ShockToElectrocuted --> ElectrocutedFinal
rules[Status.ElectrocutedBase].Feeds(Status.ElectrocutedFinal);
// ElectrocutedBase --> ElectrocutedFinal
```
The first idea was to put a condition on the suppression
```
rules[Status.ElectrocutedBase].Feeds(StatusBookKeeping.SuppressShockToElectrocuted);
rules[StatusBookKeeping.SuppressShockToElectrocuted].Suppresses(s => s <= 0, StatusBookKeeping.ShockToElectrocuted);
```
Unfortunately this does not work for some reason. The condition ONLY applies when the value is greater than 0. When the value is 0, the condition is always false, even though the condition clearly states it should be true.

The other idea was to put the condition on the Feeds, just to see how that would work.
```
rules[Status.ElectrocutedBase].Feeds(1, s => s == 0, StatusBookKeeping.SuppressShockToElectrocuted);
rules[StatusBookKeeping.SuppressShockToElectrocuted].Suppresses(StatusBookKeeping.ShockToElectrocuted);
```
This should in theory add 1 when the value of ElectrocutedBase is 0. The condition probably has the same issue as above, but I found another issue.
Apparently, when the value of ElectrocutedBase is 0, the Converter cannot return anything else than 0. In this case it returns 1.
I don't know why that check is there in the StatusSystem.ValidateConverter method, but it is there.

And lastly, don't feel pressured to answer right away. I have limited time anyways, and I am just happy that you want to answer at all. I hope these findings gives some insight in how the system could be used.

**Comment by derrickcreamer** [comment]
Yes, those limitations on zero values for feeding, suppression, and converters are intentional, because of how updated values are propagated through the tracker:  It needs to start in a known-consistent state (with the tracker being zeroed out) before anything is added; if this requirement weren't in place, then the system would need several new rules _solely_ to control how the initial values are handled. (If you were, for example, to feed several 0 values into 1s, but those 1s also suppress other values, how is that resolved?)

**Comment by derrickcreamer** [comment]
@DrHeinzDoofenshmirtz I added derived statuses to version 1.1.2 - let me know how that works for you!
